### PR TITLE
planner, statistics: build the global statistics for the partition table

### PR DESCRIPTION
**Pull request by Reminiscent**
### What problem does this PR solve?

Issue Number: related: [#18551](https://github.com/pingcap/tidb/issues/18551)

Problem Summary:
Build the global statistics for the partition tables when we execute the `analyze` statement.

### What is changed and how it works?

[Proposal](https://docs.google.com/document/d/16YouNzPG8d2Cn8pbJWY1NKCUEB_uTRmlFB1k9y5oTNQ/edit) (in Chinese)

What's Changed(In the `Dynamic-Only` mode):
In the origin implementation, we build analyze task that "collect multi partitions and save as a table" mode. You can see [PR#19846](https://github.com/pingcap/tidb/pull/19846), [PR#19899](https://github.com/pingcap/tidb/pull/19899) and [PR#20271](https://github.com/pingcap/tidb/pull/20271) for more details.

In this PR, we build the `analyze` task that "collect a partition save a partition" first. And then merge the partition-level stats which belong to the same partition table to get the global-level stats.

We will use the changed variable `analyzeTableID` for more explanation. You can see this [comment](https://github.com/pingcap/tidb/pull/22472/files#r561749461) for more details.

How it Works(In the `Dynamic-Only` mode):
1. When we build the `analyze` task, we build every task for every partition. And record the table ID to which table the partition belongs to.
2. If there are some tasks related to the partition table, we will record the partition table ID and index ID.
3. We use the partition table ID to get the corresponding partition-level stats from the storage. And merge them to build the global-level stats.

### Related changes

- PR to update pingcap/docs/pingcap/docs-cn:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- Build the global statistics for the partition table

### CODE REVIEW

File: executor/builder.go [hunk 1]
```diff
@@ -2185,11 +2185,7 @@ func (b *executorBuilder) buildAnalyze(v *plannercore.Analyze) Executor {
 			if enableFastAnalyze {
 				b.buildAnalyzeFastIndex(e, task, v.Opts)
 			} else {
-				if task.TableID.StoreAsCollectID() && len(task.TableID.CollectIDs) > 1 && !task.IndexInfo.Global && !task.IndexInfo.Unique {
-					b.buildAnalyzeFastIndex(e, task, v.Opts)
-				} else {
-					e.tasks = append(e.tasks, b.buildAnalyzeIndexPushdown(task, v.Opts, autoAnalyze))
-				}
+				e.tasks = append(e.tasks, b.buildAnalyzeIndexPushdown(task, v.Opts, autoAnalyze))
```
Comment (Reminiscent) [inline executor/builder.go:2188]: TODO: Need more investigation on when to use `buildAnalyzeFastIndex`.

File: planner/core/common_plans.go [hunk 2]
```diff
@@ -761,18 +761,20 @@ type Delete struct {
 
 // AnalyzeTableID is hybrid table id used to analyze table.
 type AnalyzeTableID struct {
-	PersistID  int64
-	CollectIDs []int64
+	PersistID int64
+	// FatherID just used in the partition table.
+	// It represents the ID of the table to which the partition belongs.
+	FatherID int64
 }
```
Comment (Reminiscent) [inline planner/core/common_plans.go:768]: Give an example to make the meaning of these IDs more clear.
Version1: no global-level stats
Version2: only global-level stats
Version3(This PR): both have partition-level stats and global-level stats, and the global-level stats are merged from the partition-level stats

Now, we have table `t`(ID = 1). And the table `t` has three partitions `p1`(ID = 2), `p2`(ID = 3), `p3`(ID = 4).

In version1, the `analyze` request has three tasks: `AnalyzeTableID{PersistID = 2, CollectIDs = {2}}`, `AnalyzeTableID{PersistID = 3, CollectIDs = {3}}` , `AnalyzeTableID{PersistID = 4, CollectIDs = {4}}`.

In version2, the `analyze` request has only one task: `AnalyzeTableID{PersistID = 1, CollectIDs = {2, 3, 4}}`.

In version3, the `analyze` request has three tasks: `AnalyzeTableID{PersistID = 2, FatherID= 1}`, `AnalyzeTableID{PersistID = 3, FatherID= 1}` , `AnalyzeTableID{PersistID = 4, FatherID= 1}`.

File: executor/analyze_test.go [hunk 3]
```diff
@@ -469,8 +469,8 @@ func (s *testFastAnalyze) TestFastAnalyze(c *C) {
 	tk.MustExec(`insert into t3 values(1, 1), (2, 2), (5, 1), (9, 3), (13, 3), (17, 5), (3, 0)`)
 	tk.MustExec(`analyze table t3`)
 	tk.MustQuery(`explain select v from t3 partition(p1) where v = 3`).Check(testkit.Rows(
-		"IndexReader_8 2.00 root  index:IndexRangeScan_7",
-		"└─IndexRangeScan_7 2.00 cop[tikv] table:t3, partition:p1, index:k(v) range:[3,3], keep order:false",
+		"IndexReader_8 2.80 root  index:IndexRangeScan_7",
+		"└─IndexRangeScan_7 2.80 cop[tikv] table:t3, partition:p1, index:k(v) range:[3,3], keep order:false",
```
Comment (Reminiscent) [inline executor/analyze_test.go:473]: TODO: Need more investigation on is it possible to use partition-level stats when there is only a single partition in where the condition.

**Comment by Reminiscent** [comment]
The reason why the test fails now is that the merged function of other statistical information like histogram, topN and so on has not been completed, we need to wait until they are all completed.

File: statistics/table.go [hunk 4]
```diff
@@ -197,6 +197,24 @@ func (t *Table) ColumnByName(colName string) *Column {
...
+}
+
+// GetStatsInfo returns their statistics according to the ID of the column or index, including histogram, CMSketch and TopN.
+func (t *Table) GetStatsInfo(ID int64, isIndex int) (*Histogram, *CMSketch, *TopN) {
+	if isIndex == 0 {
+		return t.getColumnStatsInfo(ID)
+	}
+	return t.getIndexStatsInfo(ID)
+}
+
```
Comment (qw4990) [inline statistics/table.go:217]: `getColumnStatsInfo` and `getIndexStatsInfo` are simple and not used by other functions directly, so how about merging these 3 functions into 1?

File: statistics/handle/handle.go [hunk 5]
```diff
@@ -221,6 +221,155 @@ func (h *Handle) Update(is infoschema.InfoSchema) error {
...
+	for _, partitionID := range partitionIDs {
+		h.mu.Lock()
+		partitionTable, ok := h.getTableByPhysicalID(is, partitionID)
+		h.mu.Unlock()
+		if !ok {
+			logutil.BgLogger().Debug("unknown physical ID in stats meta table, maybe it has been dropped", zap.Int64("ID", partitionID))
+			return
+		}
+		tableInfo := partitionTable.Meta()
+		partitionStats, err := h.tableStatsFromStorage(tableInfo, partitionID, false, nil)
```
Comment (qw4990) [inline statistics/handle/handle.go:324]: Should we load it from cache first?

File: statistics/handle/handle.go [hunk 6]
```diff
@@ -221,6 +221,155 @@ func (h *Handle) Update(is infoschema.InfoSchema) error {
...
+// It will both store the column stats and index stats.
+// In the column statistics, the variable `num` is equal to the number of columns in the partition table.
+// In the index statistics, the variable `num` is always equal to one.
+type GlobalStats struct {
+	num   int
+	count int64
+	hg    []*statistics.Histogram
+	cms   []*statistics.CMSketch
+	topN  []*statistics.TopN
+}
```
Comment (qw4990) [inline statistics/handle/handle.go:235]: How about exposing this struct's fields and removing its methods, which may make it clearer?

File: executor/analyze.go [hunk 7]
```diff
@@ -134,6 +156,36 @@ func (e *AnalyzeExec) Next(ctx context.Context, req *chunk.Chunk) error {
...
+				globalStatsCount := globalStats.GetCount()
+				for i := 0; i < globalStatsNum; i++ {
+					hg := globalStats.GetHistogram(i)
+					cms := globalStats.GetCMSketch(i)
+					topN := globalStats.GetTopN(i)
+					if hg == nil || cms == nil || topN == nil {
+						succ = false
+						break
+					}
+					err = statsHandle.SaveStatsToStorage(info.tableID, globalStatsCount, info.isIndex, hg, cms, topN, info.statsVersion, 1)
```
Comment (qw4990) [inline executor/analyze.go:173]: Should we update the cache after saving？

**Comment by Reminiscent** [comment]
To make it easier to review, we will split this PR into multiple sub-PRs.

File: executor/analyze.go [hunk 8]
```diff
@@ -499,7 +551,7 @@ func (e *AnalyzeColumnsExec) open(ranges []*ranger.Range) error {
 
 func (e *AnalyzeColumnsExec) buildResp(ranges []*ranger.Range) (distsql.SelectResult, error) {
 	var builder distsql.RequestBuilder
-	reqBuilder := builder.SetHandleRangesForTables(e.ctx.GetSessionVars().StmtCtx, e.tableID.CollectIDs, e.handleCols != nil && !e.handleCols.IsInt(), ranges, nil)
+	reqBuilder := builder.SetHandleRangesForTables(e.ctx.GetSessionVars().StmtCtx, []int64{e.tableID.FatherID}, e.handleCols != nil && !e.handleCols.IsInt(), ranges, nil)
```
Comment (xuyifangreeneyes) [inline executor/analyze.go:554]: Why we use `FatherID` rather than `PersistID` here? I'm kind of confused. Thx~

Comment (Reminiscent) [reply executor/analyze.go:554]: Please see this [PR](https://github.com/pingcap/tidb/pull/22554). It is somewhat different from this PR. Thanks~

**Pull request closed without merging**